channeldb/payment_control.go

```diff
@@ -305,28 +305,28 @@ func (p *PaymentControl) RegisterAttempt(paymentHash lntypes.Hash,
 			return err
 		}
 
-		p, err := fetchPayment(bucket)
+		payment, err = fetchPayment(bucket)
```

Comment:
regarding `Batch`, does it mean we also need to have an `updateErr` here? cc @joostjager

Reply:
I think it depends on the error. IIRC for batch you should only return an error if there was a problem with the database, typically some locking conflict. For something like payment not found, I think a plain `nil` must be returned.

Reply:
hmm then something is wrong here. Will double check on `Batch` and fix it in a new PR if necessary.

Reply:
wrote a test [here](https://gist.github.com/yyforyongyu/fc82b46314d3ee7128dd8312182de785), I don't think returning an error matters here. Also from [the docs](https://github.com/etcd-io/bbolt#batch-read-write-transactions):

> The trade-off is that Batch can call the given function multiple times, if parts of the transaction fail.

Reply:
So what I think happens if you return an error is that bbolt will retry the transaction. That is a useless retry if the error wasn't an internal db error. For example 'payment not found' won't magically find a payment in a retry. 

The error will be returned, but it might be less efficient because of the useless retry. I think the test doesn't surface this. You may want to add a log line inside the actual transaction to visualize duplicate calls.

Reply:
Updated the test and you are exactly right! Part of the results,
```
    channel_test.go:1599: total records: 100000
    channel_test.go:1600: total errors: 10
    channel_test.go:1601: batch updates: 104537
    channel_test.go:1602: updates with error: 20
```

Though only 100,000 records, the batch actually ran 104,537 times, whereas in the other case, it ran exactly 100,000 times. Also the error tx was retried twice.